vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom

```diff
@@ -0,0 +1,40 @@
+module ledger.endpoints.mojom;
```

Comment:
I think this file needs to be in the same folder as `ledger.mojom`.

Do we need the nested namespace here? I think `bat-native-ledger` in general went overboard with nested namespacing; all these names should be unique anyway, and it just creates more typing. I think there's a Chromium [style guide](https://chromium.googlesource.com/chromium/src.git/+/HEAD/docs/security/mojo.md#naming-conventions) for this as well.

Reply:
I think it's a good idea to keep `error_types.mojom` as close to where it belongs as possible, as opposed to putting it in `vendor/bat-native-ledger/include/bat/ledger/public/interfaces`.

Same is true for the error types in the nested `endpoints` namespace.
Also, the compile-time machinery that checks if a certain `enum class` (and its `FAILED_TO_CREATE_REQUEST` enumerator) exists relies on some basic code generation/macro expansion, e.g.:
```
RESULT_FOR(PatchWallets, void);
```
I wanted this to be put right above the declaration of the endpoint class (`PatchWallets`, in this case). `RESULT_FOR()` opens up the `mojom` namespace and forward declares the expected `enum class` as part of the check. If the error types are defined in `ledger::mojom`, I have to start the macro definition with a closing brace (since we're in the `endpoints` namespace), which I didn't like (and neither did the linter 🙂).

By the way: we wouldn't have this problem, if we didn't adhere to Chromium's naming convention (that is, to put every Mojo type in a nested `mojom` namespace), which is just a guideline to avoid name collisions (and which they even proposed to get rid of in 2019). I'm just not sure if we really want to adopt this otherwise completely useless convention?

Reply:
Is there a way to do this without the macro? I would prefer that we not use macros to declare things if possible. Can the endpoint just define an alias for the Error type?

Reply:
Yeah, I personally hate macros, but I can't think of a cleaner way of making sure:
- the error type in `error_types.mojom` is called `[Endpoint]Error`
- `[Endpoint]Error` has the enumerator `kFailedToCreateRequest`
- the endpoint implementor doesn't have to provide a forward declaration of their endpoint class, and an explicit specialization of the traits class:
  ```
  class Endpoint;

  template <>
  struct ResultFor<Endpoint> {
    using Result = base::expected<value_type, EndpointError>;
  };
  ```

Reply:
Can we drop `ResultFor<X>` and just do something like:

```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Result = base::expected<std::string, mojom::MyEndpointError>;
};
```

The compiler will catch the first two items, IIUC.

Or I suppose it could be

```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Value = std::string;
  using Error = mojom::MyEndpointError;
};
```

And still let `ResponseHandler` define the `Result` type.

Reply:
Here:
```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Result = base::expected<std::string, mojom::MyEndpointError>;
};
```
it's up to the implementor what they want the typedef to be called (e.g. `NotResult`), but certain parts of the Request/Response API (e.g. `RequestFor<>`) rely on this typedef being called `Result` (that's why I want it to be defined in `ResponseHandler<>`). As for the first two items, there's nothing that would prevent the implementor from naming the error type however they like, also, the presence of `kFailedToCreateRequest` is not enforced (other than a cryptic compilation error).

Regarding this one:
```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Value = std::string;
  using Error = mojom::MyEndpointError;
};
```
It doesn't work because of the reasons outlined [here](https://github.com/brave/brave-core/pull/15101#discussion_r972252048).

Reply:
Get rid of `ResponseHandler` then? It feels like the design space is really tightly constrained.

Reply:
The point of this design would be to enforce endpoint implementors to follow certain guidelines (and to not have any means of diverging from them), while requiring writing the least amount of code to define an endpoint. As a consequence, it doesn't leave too much design space, but this is to end up with endpoint implementations that resemble each other as much as possible. This adheres to Scott Meyers' `Make interfaces easy to use correctly and hard to use incorrectly.` advice. 🙂 Removing `ResponseHandler<>` would pretty much defeat the whole purpose.

Reply:
Fixed in 0a230cd08129036ba2e22092cb89335a75897b35 (as per our Slack convo).

Reply:
I personally (for whatever that is worth 😂) preferred the `RESULT_FOR` macro, since it encapsulates things a bit more (ex: makes its own enum, defines the class). But this edit looks OK too (fairly minimal since we only have 3 implementations here)

I definitely do understand how the macro can feel dirty; it reminds me of MFC a little bit (on older MSVC++ projects). But the usage was concise 👍  Glad you all were able to sort it out though 😄